[REVIEW] Fix ORC reader issue with decimal type

rgsl888prabhu opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
Number of values being read was't considering last update while breaking, which caused the subsequent calls behave oddly, leading to improper access of memory segments.

closes #5892

GPUtester commented:
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

rgsl888prabhu commented:
@harrism should we push this 0.16 as this is priority-0 and and change is just one line.

kkraus14 commented:
@rgsl888prabhu is there a small test file we could add to cover the error case here?

rgsl888prabhu commented:
> @rgsl888prabhu is there a small test file we could add to cover the error case here?

It is a 64KB file with which we can reproduce, would it be fine to add it ?

vuule commented:
> It is a 64KB file with which we can reproduce, would it be fine to add it ?

Is it possible to make a smaller repro?

rgsl888prabhu commented:
> > It is a 64KB file with which we can reproduce, would it be fine to add it ?
> 
> Is it possible to make a smaller repro?
> Edit: 64K not that large, but it looks like this issue should be reproducible with a much smaller file.

I am trying to create smaller file to test it.

OlivierNV commented:
Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?

rgsl888prabhu commented:
> Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).

```
if (t == 0) {
    uint32_t maxpos  = min(bs->len, bs->pos + (BYTESTREAM_BFRSZ - 8u));
    uint32_t lastpos = bs->pos;
    uint32_t n;
    for (n = 0; n < numvals; n++) {
      uint32_t pos                  = lastpos;
      *(volatile int32_t *)&vals[n] = lastpos;
      pos += varint_length<uint4>(bs, pos);
      if (pos > maxpos) break;
      lastpos = pos;
    }
    scratch->num_vals = n;
    bytestream_flush_bytes(bs, lastpos - bs->pos);
  }
```

This issue occurs when numvals=1024, so for n=1023,  pos > max_pos and it would break. But at n=1023, we have already update the value val[n] with last_pos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.

rgsl888prabhu commented:
> Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).

@OlivierNV  And as you suggested, changing the `max run = NTHREADS/2 ` also fixes the issue, but that creates issue in other scenarios as you have already mentioned.

OlivierNV commented:
> This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.

It's storing the last offset but not using it (which is likely problem since that same location is initially used to store the scale).
What about moving the line `*(volatile int32_t *)&vals[n] = lastpos;` after `if (pos > maxpos) break;` ?

rgsl888prabhu commented:
> > Hmm. That doesn't look right: n is the number of valid runs. It breaks out of the loop without incrementing it if the payload for this run would exceed the amount of data available in the buffer. What are the values in the failure case ?
> > [Edit] I think we might need to limit the max # of scale values in RLE stream1 such that it can't overflow when decoding the secondary base128 stream, eg: 512-limit for max run = NTHREADS/2 at line 1551 (might exceed the max internal buffer size, and I don't think that's a case where overflow can be properly handled because of cross-stream dependency).
> > Long-term, this kernel should really be rewritten with 512-threads with separate queues for primary & secondary stream as it would simplify things a lot and likely better for perf when index is available (it's a big change though).
> 
> @OlivierNV
> 
> ```
> if (t == 0) {
>     uint32_t maxpos  = min(bs->len, bs->pos + (BYTESTREAM_BFRSZ - 8u));
>     uint32_t lastpos = bs->pos;
>     uint32_t n;
>     for (n = 0; n < numvals; n++) {
>       uint32_t pos                  = lastpos;
>       *(volatile int32_t *)&vals[n] = lastpos;
>       pos += varint_length<uint4>(bs, pos);
>       if (pos > maxpos) break;
>       lastpos = pos;
>     }
>     scratch->num_vals = n;
>     bytestream_flush_bytes(bs, lastpos - bs->pos);
>   }
> ```
> 
> This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.



> > This issue occurs when numvals=1024, so for n=1023, pos > max_pos and it would break. But at n=1023, we have already updated the value val[n] with lastpos, and as this is using zero indexing, number of run would always be `n+1` if I am not wrong.
> 
> It's storing the last offset but not using it (which is likely problem since that same location is initially used to store the scale).
> What about moving the line `*(volatile int32_t *)&vals[n] = lastpos;` after `if (pos > maxpos) break;` ?

It still causes same issue if I move the line after break.

OlivierNV commented:
Yeah, it's not handling the buffering of scale values for decimals, and the buffer capacity is only sized for 512x64bit values, so will not handle 512 worst-case 128b decimals in one-shot consumption.
Along with moving that line after the break, it's missing code at around line 1811 (with the comment "buffer secondary stream values") to deal with buffering the leftover decimal scale values for the next pass (only done for BOOLEAN and TIMESTAMP types). The leftover values need to be stored back to the beginning of the buffer and buffered_count adjusted accordingly.
Also, it needs some added logic around line 1650 to deal with the buffered values, since in the case of decimals, we're really buffering data from the primary data stream (not a trivial change).
As a short-term WAR, we could limit the initial value of max_vals to 512 for DECIMAL type (would work here), but it's not a good long-term solution since even 512 decimal values could overflow the stream buffer (worst case is (128+6)/7 bytes per value)

OlivierNV commented:
Or alternatively to the cross-iteration buffering, a simpler solution might be to add a loop within Decode_Decimals and refill the buffer on the fly (likely less ugly than the buffering).

rgsl888prabhu commented:
Don't merge this, still under progress !!!!!!!!!!!!!!!!!!!!!!!!!

kkraus14 reviewed (changes requested):
blocking merging

rgsl888prabhu commented:
github is experiencing issues, and that's why my commit is not yet being shown in this PR.

OlivierNV commented on cpp/src/io/orc/stripe_data.cu:
```diff
@@ -1033,65 +1033,80 @@ static __device__ int Decode_Decimals(orc_bytestream_s *bs,
+    __syncthreads();
+    if (num_vals_read != numvals) {
+      bytestream_fill(bs, t);
+      if (t == 0) {
```
Needs a `__syncthreads()` between `bytestream_fill()` and `if (t == 0) {` (race condition with `bs->fill_count`)

OlivierNV reviewed (approved):
LGTM

rgsl888prabhu commented:
Please select `Hide whitespace changes` option as normal diff shows lot of unrelated things.

kkraus14 commented:
@rgsl888prabhu did you forget to check in the test files?

vuule commented:
> @rgsl888prabhu did you forget to check in the test files?

I see the Python test file change

kkraus14 commented:
The Python test file refers to: `TestOrcFile.decimal.same.values.orc` and `TestOrcFile.decimal.multiple.values.orc` which aren't checked into CI as far as I could tell

rgsl888prabhu commented:
> @rgsl888prabhu did you forget to check in the test files?

@vuule He meant orc format files for tests.

@kkraus14  I just uploaded them.

kkraus14 commented:
@vuule do you want to review before we mark this as ready to merge?

vuule commented:
@rgsl888prabhu Do you think we should check for perf regression before merging this? Looks unlikely to me, but I don't know the context well.

vuule reviewed (approved):
Good stuff. Thanks for finding smaller repro files.

rgsl888prabhu commented:
> Good stuff. Thanks for finding smaller repro files.

Thanks to @OlivierNV.

rgsl888prabhu commented:
> @rgsl888prabhu Do you think we should check for perf regression before merging this? Looks unlikely to me, but I don't know the context well.

Will try it and let you know.

rgsl888prabhu commented:
Tried with a ORC dataset with just one column with `440000000`, orc file size is 250 MB.

There is not much change in spped.
older
```
In [2]: %timeit cudf.read_orc("/home/rgsl888/orc_8GB_250MB.orc")
1.7 s ± 6.05 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```
In [2]: %timeit cudf.read_orc("/home/rgsl888/orc_8GB_250MB.orc")
1.64 s ± 260 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
